### PR TITLE
Sort search results by download count.

### PR DESCRIPTION
**Pull request by mhilton**
Search results can be sorted by download
counts using sort=[-]downloads.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/623/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/624/

**Comment by mhilton** [comment]
To QA: run charmd and download some archives, perform a search including those archives. The following command is helpful.

```
http :8080/v4/search?sort=-downloads&include=stats
```

File: internal/charmstore/search.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func (s *Store) UpdateSearch(r *charm.Reference) error {
 		return nil
 	}
 	var entity mongodoc.Entity
-	if err := s.DB.Entities().Find(bson.M{"user": r.User, "name": r.Name, "series": r.Series}).Sort("-Revision").One(&entity); err != nil {
+	if err := s.DB.Entities().Find(bson.M{"user": r.User, "name": r.Name, "series": r.Series}).Sort("-revision").One(&entity); err != nil {
```
Comment (rogpeppe) [inline internal/charmstore/search.go:80]: This looks like a previous bug. Is there a test that tests this bug specifically?

File: internal/charmstore/stats.go [hunk 2]
```diff
@@ -520,3 +520,29 @@ func (s *Store) aggregateStats(key []string, prefix bool) (AggregatedCounts, err
...
+}
+
+// IncrementDownloadCounts updates the download statistics for charm id in both
+// the statistics database and the search database.
+func (s *Store) IncrementDownloadCounts(id *charm.Reference) error {
+	key := EntityStatsKey(id, params.StatsArchiveDownload)
+	if err := s.IncCounter(key); err != nil {
+		return errgo.Notef(err, "cannot increase stats counter for %v", key)
+	}
+	// TODO(mhilton) when this charmstore is being used by juju find a more
```
Comment (rogpeppe) [inline internal/charmstore/stats.go]: s/juju find/juju, find/

File: internal/charmstore/search.go (hunk 1 above)
Comment (mhilton) [inline internal/charmstore/search.go:80]: There wasn't I've changed the test that checks this functionality to a form that would exercise it.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/629/

File: internal/charmstore/stats.go (hunk 2 above)
Comment (mhilton) [inline internal/charmstore/stats.go]: done

File: internal/charmstore/stats.go [hunk 3]
```diff
@@ -520,3 +520,29 @@ func (s *Store) aggregateStats(key []string, prefix bool) (AggregatedCounts, err
 	}
 	return counts, nil
 }
+
+// IncrementDownloadCounts updates the download statistics for charm id in both
```
Comment (frankban) [inline internal/charmstore/stats.go]: for entity id?

File: internal/charmstore/stats.go [hunk 4]
```diff
@@ -520,3 +520,29 @@ func (s *Store) aggregateStats(key []string, prefix bool) (AggregatedCounts, err
...
+// background using a separate goroutine.
+func (s *Store) IncrementDownloadCountsAsync(id *charm.Reference) {
+	go func() {
+		if err := s.IncrementDownloadCounts(id); err != nil {
+			logger.Errorf("cannot increase download counter for %v: %s", id, err)
+		}
+	}()
+}
+
+// IncrementDownloadCounts updates the download statistics for charm id in both
```
Comment (frankban) [inline internal/charmstore/stats.go]: ditto

**Comment by frankban** [comment]
:+1:  and QA OK.
Thank you!

**Comment by rogpeppe** [comment]
LGTM

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/632/

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**